helix-term/src/ui/editor.rs

```diff
@@ -488,22 +525,27 @@ impl EditorView {
                     use helix_core::graphemes::{grapheme_width, RopeGraphemes};
 
                     for grapheme in RopeGraphemes::new(text) {
-                        let out_of_bounds = visual_x < offset.col as u16
-                            || visual_x >= viewport.width + offset.col as u16;
-
                         if LineEnding::from_rope_slice(&grapheme).is_some() {
-                            if !out_of_bounds {
+                            if !out_of_bounds(visual_x) {
                                 // we still want to render an empty cell with the style
                                 surface.set_string(
                                     viewport.x + visual_x - offset.col as u16,
                                     viewport.y + line,
                                     &newline,
                                     style.patch(whitespace_style),
                                 );
+                                visual_x += 1;
```

Comment:
Why was this not necessary before, but is now?

Reply:
I believe because it is due to the addition of virtual text.

But this `+= 1` does not seemed to support double width characters like CJK, otherwise I think it shouldn't always `+= 1` and sometimes it should `+= 2`.

Reply:
I think this works as intended in all cases.

`visual_x` point is placed **after** the last character (always incremented by the correct `grapheme` width in the else clause). The `visual_x += 1` only moves the virtual text to the right by one cell so there is space between the annotation and the content of the line. It might be nice to configure the amount of space in the future but I think that can be done in a followup PR.

This was not required before this PR because `visual_x` is reset back to `0` right after this condition (as a new line starts) on `master`. However this PR still requires the `visual_x` (plus the space) to render the EOL annotations.